api/src/resources/audit.rs

```diff
@@ -0,0 +1,209 @@
+    event_id: AuditEventId,
+    event_type: AuditEventType,
+    #[serde(skip_serializing_if = "Vec::is_empty", default)]
+    project_ids: Vec<ProjectId>,
+    tenant_ids: Vec<AuditTenantId>,
+    timestamp: DateTime<Utc>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct PrintableAuditEvent {
+    pub actor_email: Email,
+    pub actor_tenant_name: AuditTenantName,
+    pub event_type: AuditEventType,
+    pub dataset_names: Vec<DatasetName>,
+    pub event_id: AuditEventId,
+    pub project_names: Vec<ProjectName>,
+    pub tenant_names: Vec<AuditTenantName>,
+    pub timestamp: DateTime<Utc>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct AuditDataset {
+    id: DatasetId,
+    name: DatasetName,
+    project_id: ProjectId,
+    title: String,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct AuditProject {
+    id: ProjectId,
+    name: ProjectName,
+    tenant_id: AuditTenantId,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct AuditTenant {
+    id: AuditTenantId,
+    name: AuditTenantName,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct AuditUser {
+    display_name: Username,
+    email: Email,
+    id: UserId,
+    tenant_id: AuditTenantId,
+    username: Username,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct AuditQueryResponse {
+    audit_events: Vec<AuditEvent>,
+    projects: Vec<AuditProject>,
+    pub continuation: Option<Continuation>,
+    datasets: Vec<AuditDataset>,
+    tenants: Vec<AuditTenant>,
+    users: Vec<AuditUser>,
+}
+
+impl AuditQueryResponse {
+    pub fn into_iter_printable(self) -> QueryResponseIterator {
+        QueryResponseIterator {
+            response: self,
+            index: 0,
+        }
+    }
+
+    fn get_user(&self, user_id: &UserId) -> Option<&AuditUser> {
+        self.users.iter().find(|user| user.id == *user_id)
+    }
+
+    fn get_dataset(&self, dataset_id: &DatasetId) -> Option<&AuditDataset> {
+        self.datasets
+            .iter()
+            .find(|dataset| dataset.id == *dataset_id)
+    }
+
+    fn get_project(&self, project_id: &ProjectId) -> Option<&AuditProject> {
+        self.projects
+            .iter()
+            .find(|project| project.id == *project_id)
+    }
+
+    fn get_tenant(&self, tenant_id: &AuditTenantId) -> Option<&AuditTenant> {
+        self.tenants.iter().find(|tenant| tenant.id == *tenant_id)
+    }
+}
+
+pub struct QueryResponseIterator {
+    response: AuditQueryResponse,
+    index: usize,
+}
+
+impl Iterator for QueryResponseIterator {
+    type Item = PrintableAuditEvent;
+    fn next(&mut self) -> Option<Self::Item> {
+        let response = self.response.audit_events.get(self.index).map(|event| {
+            let actor_email = &self
+                .response
+                .get_user(&event.actor_user_id)
+                .unwrap_or_else(|| panic!("Could not find user for id `{}`", event.actor_user_id.0))
+                .email;
```

Comment:
If perf is ever an issue here, I would suggest converting the raw `response` into some hashmap lookups for users/datasets etc. at the point you convert it to `QueryResponseIterator`, so you're not repeatedly scanning with `find`.

But happy this works.

Note that I think it is technically possible for us to not return all linked resources (e.g. if the user has been deleted, won't be included).

I'm happy for your cases this is not likely, but if you were pushing this out to other users, I'd want some nicer error handling here.